core/src/replay_stage.rs

```diff
@@ -645,71 +641,103 @@ impl ReplayStage {
     }
 
     #[allow(clippy::type_complexity)]
-    fn generate_votable_banks(
+    fn select_fork(
         ancestors: &HashMap<u64, HashSet<u64>>,
         bank_forks: &Arc<RwLock<BankForks>>,
         tower: &Tower,
         progress: &mut HashMap<u64, ForkProgress>,
-    ) -> Vec<(u128, Arc<Bank>, HashMap<u64, StakeLockout>, u64)> {
+    ) -> Option<(Arc<Bank>, u64)> {
         let tower_start = Instant::now();
         let frozen_banks = bank_forks.read().unwrap().frozen_banks();
+
         trace!("frozen_banks {}", frozen_banks.len());
         let mut votable: Vec<(u128, Arc<Bank>, HashMap<u64, StakeLockout>, u64)> = frozen_banks
             .values()
-            .filter(|b| {
-                let has_voted = tower.has_voted(b.slot());
-                trace!("bank has_voted: {} {}", b.slot(), has_voted);
+            .filter(|bank| {
+                let has_voted = tower.has_voted(bank.slot());
+                trace!("bank has_voted: {} {}", bank.slot(), has_voted);
                 !has_voted
             })
-            .filter(|b| {
-                let is_locked_out = tower.is_locked_out(b.slot(), &ancestors);
-                trace!("bank is is_locked_out: {} {}", b.slot(), is_locked_out);
-                !is_locked_out
-            })
             .map(|bank| {
+                let (stake_lockouts, total_staked) = tower.collect_vote_lockouts(
+                    bank.slot(),
+                    bank.vote_accounts().into_iter(),
+                    &ancestors,
+                );
+                Self::confirm_forks(tower, &stake_lockouts, total_staked, progress, bank_forks);
                 (
-                    bank,
-                    tower.collect_vote_lockouts(
-                        bank.slot(),
-                        bank.vote_accounts().into_iter(),
-                        &ancestors,
-                    ),
+                    bank.clone(),
+                    stake_lockouts,
+                    total_staked,
                 )
             })
-            .filter(|(b, (stake_lockouts, total_staked))| {
+            .filter(|(bank, _, _)| {
+                let last_slot = *tower.last_vote().slots.last().unwrap_or(&0);
+                let first_slot = *tower.last_vote().slots.first().unwrap_or(&0);
+                assert!(first_slot <= last_slot);
+                let new_bank = bank.slot() > last_slot;
```

Comment:
i think there's still an issue if the heaviest bank is:

1) Greater than the last voted slot
2) Validators are locked out on that heaviest bank:
3) That heaviest bank has no descendants

I imagine the case with 3 validators, with A: 24%, B: 24%, C: 24%, D: 28% of the stake, where 

A is the leader for slots 1-4, B: 4-8, C: 9-10, and D 11-14. If they each create their own forks then vote on their own forks, D will have the heaviest fork. If D then dies, none of the other validators will vote and make any progress.